kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/OpenIDConnectionUtilsBehaviorTest.java

```diff
@@ -0,0 +1,372 @@
+            "}");
+      }
+
+      @Nested
+      @DisplayName("With 404 token response")
+      class WithNotFoundTokenResponse {
+
+        private String result;
+
+        @BeforeEach
+        void setUp() throws Exception {
+          httpClientFactory.expect("/token", 404, "Not Found /token");
+          result = resolveOIDCTokenFromAuthConfig(originalConfig, authProviderConfig, httpClientBuilder)
+              .get(10, TimeUnit.SECONDS);
+        }
+
+        @Test
+        @DisplayName("Resolves token from auth provider config (fallback)")
+        void fallbacksToOriginalToken() {
+          assertThat(result).isEqualTo("original-token");
+        }
+
+        @Test
+        @DisplayName("Logs refresh token response")
+        void logsRefreshTokenResponse() {
+          assertThat(systemErr.toString()).contains("Response: Not Found /token");
+        }
+
+        @Test
+        @DisplayName("Logs token fallback warning")
+        void logsTokenFallbackWarning() {
+          assertThat(systemErr.toString())
+              .contains(
+                  "token response did not contain an id_token, either the scope \\\"openid\\\" wasn't requested upon login, or the provider doesn't support id_tokens as part of the refresh response.");
+        }
+      }
+
+      @Nested
+      @DisplayName("With invalid token response body")
+      class WithInvalidTokenResponseBody {
+
+        private String result;
+
+        @BeforeEach
+        void setUp() throws Exception {
+          httpClientFactory.expect("/token", 200, "Not JSON");
+          result = resolveOIDCTokenFromAuthConfig(originalConfig, authProviderConfig, httpClientBuilder)
+              .get(10, TimeUnit.SECONDS);
+        }
+
+        @Test
+        @DisplayName("Resolves token from auth provider config (fallback)")
+        void fallbacksToOriginalToken() {
+          assertThat(result).isEqualTo("original-token");
+        }
+
+        @Test
+        @DisplayName("Logs JSON parsing error")
+        void logsJsonParsingError() {
+          assertThat(systemErr.toString())
+              .contains("Failure in fetching refresh token:")
+              .contains("Cannot construct instance of `java.util.LinkedHashMap`");
+        }
+
+        @Test
+        @DisplayName("Logs token fallback warning")
+        void logsTokenFallbackWarning() {
+          assertThat(systemErr.toString())
+              .contains(
+                  "token response did not contain an id_token, either the scope \\\"openid\\\" wasn't requested upon login, or the provider doesn't support id_tokens as part of the refresh response.");
+        }
+      }
+
+      @Test
+      @DisplayName("With valid token repsonse and missing kube config, logs warning")
+      void withValidTokenResponseAndMissingKubeConfig() throws Exception {
+        Files.delete(originalConfig.getFile().toPath());
+        httpClientFactory.expect("/token", 200, "{" +
+            "\"id_token\": \"new-token\"" +
```

Comment:
In token response, both access token and refresh token are returned. I think we should also verify that refresh token is also updated/persisted after we process a valid token response.

Reply:
Fixed, is this what you meant, or should I add anything else?